### ISSUE TITLE
Routing issue.

### ISSUE THREAD

**Issue by Algoclogic** [daveh/php-mvc#22]
Hi,
I tried to add other routes to the different links. I am not able to understand how routing working in you package.

Kindly reply.
Thanks

**Comment by cagcak** [comment]
if everything works normally after fresh install, follow the steps below:
- Register controller: Creating a new php file inside App/Controllers/ folder, or modifying existing
- Add method: Prepare method to call action. For contactUs, create contactUsAction ref: https://github.com/daveh/php-mvc/blob/dc047dfa7a5b99ceadcc18c4b3f31e5e1a45cd37/Core/Controller.php#L44 
- Then create the view file inside Views folder. Reference to render in the method you've just created like this:
 `View::render('contact/newViewFile.php');`
or vith the power of twig template engine:
`View::renderTemplate('contact/newViewFile.twig');`
- Finally add or register for router inside frontend controller (public/index.php) like below:
`$router->add('_SLUG_', ['controller' => '_CONTROLLER_', 'action' => '_METHOD-NAME_']);`
in this case: 
`$router->add('contact-us', ['controller' => 'Home', 'action' => 'contactUs']);`

**Comment by daveh** [comment]
You have to add routes in the public/index.php file that match the URLs that you want to controllers and actions. Check out the documentation about [routing](https://github.com/daveh/php-mvc#routing) in the README file. If you still have problems, please supply your list of routes so I can see where the problem might be.

**Comment by Algoclogic** [comment]
Thanks, guys, I had omitted a dispatch statement in routes, because of which I was not able to hit other routes.
Now working fine.

It would be great if you mention some Model-based DB interaction. Like laravel. 
I believe this package has that stuff as well.

Thanks.

**Comment by daveh** [comment]
DB interaction code is covered on [the course](http://daveh.io/r/ghmvc) this framework results from.

I hope that helps!

Cheers,
Dave

**Comment by cagcak** [comment]
Hi @daveh . 
One last thing to clarify for routing issue; 
if I add:
`$router->add('products', ['controller' => 'Products', 'action' => 'index']);`
to the routing table, uri perfectly matches. But if I try `products/ `I get error which no 'route matches'.
if add also `$router->add('products/', ['controller' => 'Products', 'action' => 'index']);` it solves the problem.
I think it is pointless to do that for every route, if any suggestion would be nice for handling this problem.
Thanks in advance.

**Comment by daveh** [comment]
Yes, you're right, there's no default action in the router. If you have generic routes like this:

```
$router->add('{controller}/{action}');
$router->add('{controller}', ['action' => 'index']);
```

then you only need the second route to make the index action the default.

I plan to add a default action to the router in the near future when I get time.

Does that help?

**Comment by cagcak** [comment]
Hi @daveh .
I understand the requirement now. Actually the problem I am trying to solve is the same with #1 . 
I prepared a method like below:
```
    public function editAction()
    {
      $id = $this->route_params['id'];
      . . .
```
in the route table: 
```
. . .
$router->add('{controller}/{action}');
$router->add('{controller}/{action}/{id:\d+}');
$router->add('products/', ['controller' => 'Products', 'action' => 'index']);
. . .
```
I aspect the url like `localhost/php-mvc/products/12` works. Instead it results `Undefined index: id`
Additionally there is no any problem in data, I tested just to be sure like:
 ```
     public function editAction($id = 12)
     {
         . . .
```
Any suggestion would be great. Thanks in advance. 

> EDIT: the above url is misspelled, actual issued url:  `localhost/php-mvc/products/edit/12`

**Comment by daveh** [comment]
If you're running the framework inside a subfolder as you seem to be, then it looks like the URL `localhost/php-mvc/products/12` won't match any of those routes. You'd need a route like this:

`$router->add('{controller}/{id:\d+}', ['action' => 'edit']);`

or alternatively one of your existing routes:

`$router->add('{controller}/{action}/{id:\d+}');`

and a URL like this: `localhost/php-mvc/products/12/edit`

As for getting the value of the id variable inside the action method, it doesn't work that way, as arguments to the action methods - instead you can access it in the `route_params` property of the controller, for example:

```
public function editAction()
{
    $id = $this->route_params['id'];
    ...
```

**Comment by cagcak** [comment]
Hi @daveh .
Sorry for spelling mistake in the url. It is actually:

`localhost/php-mvc/products/edit/12`

I did try some modifications and your suggestions like below:

`$router->add('{controller}/{action}/{id:\d+}', ['action' => 'edit']);`

but no luck. Still getting `undefined index : id` error 

My Products controller has exactly the same thing like you have told:
```
public function editAction()
{
    $id = $this->route_params['id'];
    ...
```
TEST: by the `localhost/php-mvc/products/edit/2` url `var_dump(route_params['id'])` results as below: 

`Message: 'Use of undefined constant route_params - assumed 'route_params''`

OK, I was trying to pass parameter from request url to method of course it didn't work. Else I just add the following function to the controller:
```
    public static function get_params_id()
    {
      $req = $_SERVER['REQUEST_URI'];
      $req = substr($req, strrpos($req, '/') + 1);
      return $req;
    } 
``` 
This simply returns the url `localhost/php-mvc/products/edit/2` as 2. Then I use it in method to process some jobs as:
`$this->id = static::get_params_id();`

**Comment by daveh** [comment]
I'm not sure why the parameter isn't coming from the route - what happens if you try this in the action method:

`var_dump($this->route_params);`

**Comment by cagcak** [comment]
For `var_dump($this->route_params);` inside the action method: 
It says: undefined index : id

For `var_dump(route_params);` inside the action method: 
It says: `Message: 'Use of undefined constant route_params - assumed 'route_params''`

I removed error handlers in frontend controller, still receiving same kind of errors.

**Comment by daveh** [comment]
`var_dump(route_params);` is incorrect code, it needs to be `var_dump($this->route_params);`

Please try putting this at the top of the action method:

```
var_dump($this->route_params);
exit();
```

so that this is the only code that runs. What is printed out?

**Comment by cagcak** [comment]
Hi @daveh 
Still I couldn't solve the issue so here is the result of `var_dump($this->route_params);exit();` : 

url:      `http://localhost/php-mvc/products/edit/29`
result: `array(0) { }`

Here is the controller edit action method looks like:
```
    public function editAction()
    {
      // $id = $this->route_params['id'];
      var_dump($this->route_params);
      exit();
      if ($_SERVER['REQUEST_METHOD'] == 'POST') {
        $_POST = filter_input_array(INPUT_POST, FILTER_SANITIZE_STRING);
        $data = [
          'page_title' => 'Edit product',
          'product_id' => $id,
          'product_title' => trim($_POST['pro_posted_title']),
          'product_body' => trim($_POST['pro_posted_body']),
        ];
        . . .
```
I am actually using an additional method which I wrote above `get_params_id()`, it is not efficient to use widely but it works.

**Comment by daveh** [comment]
Hi,

it looks like the `route_params` array is empty, which is odd - please can you post your routes from public/index.php?

Thanks,
Dave

**Comment by cagcak** [comment]
Hi @daveh . 
Here is the routing table from my side:

```
$router = new Core\Router();

$router->add('{controller}/{action}');
$router->add('{controller}/{action}/{id:\d+}');
// $router->add('{controller}/{action}/{id:\d+}', ['controller' => 'Products','action' => 'edit']);
$router->add('', ['controller' => 'Home', 'action' => 'index']);
$router->add('about', ['controller' => 'Home', 'action' => 'about']);
$router->add('products', ['controller' => 'Products', 'action' => 'index']);
$router->add('products/{controller}/{action}');
$router->add('products/{controller}/{action}/{id:\d+}');
$router->add('products/subproduct/', ['controller' => 'Pros\SomePro', 'action' => 'index']);

$router->dispatch($_SERVER['QUERY_STRING']);
```

Everything seems right but it doesn't work.

Best regards and
Thanks in advance.

**Comment by daveh** [comment]
It looks that that URL should be matching the second route, so the params array should contain the ID at least. Please can you post the full controller class?

**Comment by cagcak** [comment]
Hi @daveh 
I don't understand. `{controller}/{action}` works perfectly but not `{id} `. If `{id}` route has to match, adding hundreds of routes for singular products id would be pointless.

Here is the Product controller. It works for every method now with `get_params_id()` helper function currently.

> Side note: __('key') is a custom getText helper function

```
<?php

  namespace App\Controllers;

  use \Core\View;
  use \App\Models\Product;
  /**
  * Products controller
  */
  class Products extends \Core\Controller
  {

    function __construct($table)
    {
      $this->table = 'products_list';
      $this->productModel = new Product($this->table);
      $this->id = static::get_params_id();
    }

    /**
    * Products index page
    * List all products with CRUD controls
    * @return void
    */
    public function indexAction()
    {
      $pro = $this->productModel->getProductsList();

      $data = [
        'title' => __('products'),
        'products' => $pro
      ];

      View::render('products/index', $data);
    }

    /**
    * Products add action method
    * ADD: SQL INSERT
    * For CRUD control
    * @return void
    */
    public function addAction()
    {
      if ($_SERVER['REQUEST_METHOD'] == 'POST') {
        $_POST = filter_input_array(INPUT_POST, FILTER_SANITIZE_STRING);

        $data = [
          'page_title' => __('products') . ' | ' . __('pro-add'),
          'product_id' => $this->productModel->getProductListCount(),
          'product_title' => trim($_POST['pro_posted_title']),
          'product_body' => trim($_POST['pro_posted_body']),
          'product_title_error' => '',
          'product_body_error' => ''
        ];
        $data['product_id']++;

        $data = productValidate($data);
        if (empty($data['product_title_error']) && empty($data['product_body_error']) ) {
          if ($this->productModel->addProduct($data)) {
            flash('products_mvc_message',
            redirect('products');
          } else {
            die('Failed on add product');
          }
        } else {
          View::render('products/_action/add', $data);
        }
      } else {
        $data = [
          'page_title' => __('products') . ' | ' . __('pro-add'),
          'product_title' => '',
          'product_body' => '',
          'product_title_error' => '',
          'product_body_error' => ''
        ];
        View::render('products/_action/add', $data);
      }
    }

    /**
    * Products edit action method
    * EDIT: SQL INSERT
    * For CRUD control
    * @return void
    */
    public function editAction()
    {
      if ($_SERVER['REQUEST_METHOD'] == 'POST') {
        $_POST = filter_input_array(INPUT_POST, FILTER_SANITIZE_STRING);

        $data = [
          'page_title' => __('products') . ' | ' . __('edit'),
          'product_id' => $this->id,
          'product_title' => trim($_POST['pro_posted_title']),
          'product_body' => trim($_POST['pro_posted_body'])
        ];

        $data = productValidate($data);
        if (empty($data['product_title_error']) && empty($data['product_body_error'])) {
          if ($this->productModel->updateProduct($data)) {
              flash('products_mvc_message', __('pro-updated'));
              redirect('products');
          } else {
            die('Failed on edit product');
          }
        } else {
          # Show errors
          flash('products_mvc_message', $data['product_title_error'] . $data['product_body_error'], 'alert alert-danger');
          redirect('products/edit/' . $data['product_id']);
        }
      } else {
        $id = parse_url($this->id, PHP_URL_PATH);
        $product = $this->productModel->getProductById($id);

        $data = [
          'page_title' => __('products') . ' | ' . __('edit'),
          'product_current' => $product,
          'product_title' => $product->title,
          'product_body' => $product->body,
          'product_title_error' => '',
          'product_body_error' => ''
        ];
        View::render('products/_action/edit', $data);
      }
    }

    /**
    * Products show action method
    * SHOW: SQL SELECT
    * @return void
    */
    public function showAction()
    {
      $product = $this->productModel->getProductById($this->id);

      $data = [
        'current_product' => $product
      ];

      View::render('products/_action/show', $data);

    }

    /**
    * Products delete action method
    * SHOW: SQL DELETE
    * @return void
    */
    public function deleteAction()
    {
      $product = $this->productModel->getProductById($this->id);

      if ($_SERVER['REQUEST_METHOD'] == 'POST') {
        if ($this->productModel->deleteProduct($product->ID)) {
         # Deleting files and routes after successfull db operations
         delete_files(C. 'Products' .l. safeClassName($product->title));
          flash('products_mvc_message', '> '.$product->title.' | '.__('pro-deleted-all'));
          redirect('products');
        } else {
          die('error on delete data');
        }
      } else {
        redirect('products');
      }
    }
  }

```

**Comment by daveh** [comment]
I think I see the problem. When the router creates an object of the controller class, in the `dispatch` method, it passes in the parameters it's got from the route:

`$controller_object = new $controller($this->params);`

These are assigned to the route_params property of the controller, which is defined in the Core/Controller class. This is done in the constructor method of the Core/Controller class. However, in _your_ controller class, you're overriding the constructor, so the router can no longer assign the route parameters, which is why that array is empty.

If you want to add code to the constructor, you need to include the original argument, and call the parent class constructor method inside it:

`parent::__construct($route_params);`

Then you should be able to get the value of the ID from the route in the action method.

**Comment by cagcak** [comment]
Hi @daveh 
You're exactly right. I've just changed the constructor method like below: 

```
. . .
function __construct($route_params)
{
    parent::__construct($route_params);
    $this->table = 'products_list';
    . . .
. . .
```
and it worked perfectly. What a stupid mistake I did. I am so appreciated. Sorry for the headach. I think it might be better to point this detail in repo's documentation. This is the most cleaner mvc structure I've ever met.
Thanks again for the tutorial.
Best regards.
